File: ReactAndroid/src/main/java/com/facebook/react/animated/ValueAnimatedNode.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.animated;

import androidx.annotation.Nullable;
import com.facebook.react.bridge.ReadableMap;

/**
 * Basic type of animated node that maps directly from {@code Animated.Value(x)} of Animated.js
 * library.
 */
/*package*/ class ValueAnimatedNode extends AnimatedNode {
  /*package*/ double mValue = Double.NaN;
  /*package*/ double mOffset = 0;
  private @Nullable AnimatedNodeValueListener mValueListener;

  public ValueAnimatedNode() {
    // empty constructor that can be used by subclasses
  }

  public ValueAnimatedNode(ReadableMap config) {
    mValue = config.getDouble("value");
    mOffset = config.getDouble("offset");
  }

  public double getValue() {
    if (Double.isNaN(mOffset + mValue)) {
      this.update();
    }
    return mOffset + mValue;
  }

  public Object getAnimatedObject() {
    return null;
  }

  public void flattenOffset() {
    mValue += mOffset;
    mOffset = 0;
  }

  public void extractOffset() {
    mOffset += mValue;
    mValue = 0;
  }

  public void onValueUpdate() {
    if (mValueListener == null) {
      return;
    }
    mValueListener.onValueUpdate(getValue());
  }

  public void setValueListener(@Nullable AnimatedNodeValueListener listener) {
    mValueListener = listener;
  }

  public String prettyPrint() {
    return "ValueAnimatedNode[" + mTag + "]: value: " + mValue + " offset: " + mOffset;
  }
}

File: packages/react-native/temp/package/ReactAndroid/src/main/java/com/facebook/react/animated/ValueAnimatedNode.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.animated

import com.facebook.react.bridge.ReadableMap

/**
 * Basic type of animated node that maps directly from {@code Animated.Value(x)} of Animated.js
 * library.
 */
public open class ValueAnimatedNode(config: ReadableMap? = null) : AnimatedNode() {
  @JvmField internal var nodeValue: Double = config?.getDouble("value") ?: Double.NaN
  @JvmField internal var offset: Double = config?.getDouble("offset") ?: 0.0
  private var valueListener: AnimatedNodeValueListener? = null

  public fun getValue(): Double {
    if ((offset + nodeValue).isNaN()) {
      this.update()
    }
    return offset + nodeValue
  }

  public open fun getAnimatedObject(): Any? = null

  public fun flattenOffset(): Unit {
    nodeValue += offset
    offset = 0.0
  }

  public fun extractOffset(): Unit {
    offset += nodeValue
    nodeValue = 0.0
  }

  public fun onValueUpdate(): Unit {
    valueListener?.onValueUpdate(getValue())
  }

  public fun setValueListener(listener: AnimatedNodeValueListener?): Unit {
    valueListener = listener
  }

  override public fun prettyPrint(): String =
      "ValueAnimatedNode[$tag]: value: $nodeValue offset: $offset"
}
